pollux/lib/core/src/main/scala/io/iohk/atala/pollux/core/service/PresentationServiceImpl.scala

```diff
@@ -882,6 +925,87 @@ private class PresentationServiceImpl(
       PresentationRecord.ProtocolState.PresentationVerificationFailed
     )
 
+  override def verifyAnoncredPresentation(
+      presentation: Presentation,
+      requestPresentation: RequestPresentation,
+      recordId: DidCommID
+  ): ZIO[WalletAccessContext, PresentationError, PresentationRecord] = {
+    for {
+      maybeRecord <- presentationRepository
+        .getPresentationRecord(recordId)
+        .mapError(RepositoryError.apply)
+      record <- ZIO
+        .fromOption(maybeRecord)
+        .mapError(_ => RecordIdNotFound(recordId))
+      credentialsToUseJson <- ZIO
+        .fromOption(record.anoncredCredentialsToUse)
+        .mapError(_ => InvalidFlowStateError(s"No request found for this record: $recordId"))
+      anoncredCredentialProofs <-
+        AnoncredCredentialProofsV1.schemaSerDes
+          .deserialize(credentialsToUseJson)
+          .mapError(error => PresentationError.UnexpectedError(error.error))
+      requestPresentation <- ZIO
+        .fromOption(record.requestPresentationData)
+        .mapError(_ => InvalidFlowStateError(s"RequestPresentation not found: $recordId"))
+      issuedValidCredentials <- credentialRepository
+        .getValidAnoncredIssuedCredentials(
+          anoncredCredentialProofs.credentialProofs.map(_.credential).map(DidCommID(_))
+        )
+        .mapError(RepositoryError.apply)
+      schemaIds = issuedValidCredentials.flatMap(_.schemaId)
+      schemaMap <-
+        ZIO
+          .collectAll(schemaIds.map { schemaId =>
+            resolveSchema(schemaId)
+          })
+          .map(_.toMap)
+      credentialDefinitionIds = issuedValidCredentials.flatMap(_.credentialDefinitionId)
+      credentialDefinitionMap <-
+        ZIO
+          .collectAll(credentialDefinitionIds.map { credentialDefinitionId =>
+            for {
+              credentialDefinition <- credentialDefinitionService
+                .getByGUID(credentialDefinitionId)
+                .mapError(e => UnexpectedError(e.toString))
+            } yield (
+              credentialDefinition.longId,
+              AnoncredCredentialDefinition(credentialDefinition.definition.toString)
+            )
+          })
+          .map(_.toMap)
+      serialisedPresentation <- presentation.attachments.head.data match {
+        case Base64(data) => ZIO.succeed(AnoncredPresentation(new String(JBase64.getUrlDecoder.decode(data))))
+        case _            => ZIO.fail(InvalidAnoncredPresentation("Expecting Base64-encoded data"))
+      }
+      serialisedPresentationRequest <- requestPresentation.attachments.head.data match {
+        case Base64(data) => ZIO.succeed(AnoncredPresentationRequest(new String(JBase64.getUrlDecoder.decode(data))))
+        case _            => ZIO.fail(InvalidAnoncredPresentationRequest("Expecting Base64-encoded data"))
+      }
+      isValid <-
+        ZIO
+          .fromEither(
```

Comment:
probably, ZIO.fromTry can be used here without Either

Reply:
will fix in follow up